video/out/vo_kitty.c

```diff
@@ -0,0 +1,435 @@
+    bool resized = (prev_width != vo->dwidth || prev_height != vo->dheight);
+#endif
+
+    if (resized)
+        reconfig(vo, vo->params);
+
+    resized = false;
+
+    if (frame->current) {
+        mpi = mp_image_new_ref(frame->current);
+        struct mp_rect src_rc = p->src;
+        src_rc.x0 = MP_ALIGN_DOWN(src_rc.x0, mpi->fmt.align_x);
+        src_rc.y0 = MP_ALIGN_DOWN(src_rc.y0, mpi->fmt.align_y);
+        mp_image_crop_rc(mpi, src_rc);
+
+        mp_sws_scale(p->sws, p->frame, mpi);
+    } else {
+        mp_image_clear(p->frame, 0, 0, p->width, p->height);
+    }
+
+    struct mp_osd_res res = { .w = p->width, .h = p->height };
+    osd_draw_on_image(vo->osd, res, mpi ? mpi->pts : 0, 0, p->frame);
+
+#if HAVE_POSIX
+    if (p->opts.use_shm) {
+        p->shm_fd = shm_open(p->shm_path, O_CREAT | O_RDWR, S_IRUSR | S_IWUSR);
+        if (p->shm_fd == -1) {
+            MP_WARN(vo, "Failed to create shared memory object");
+            return;
+        }
+
+        if (ftruncate(p->shm_fd, p->buffer_size) == -1) {
+            MP_WARN(vo, "Failed to truncate shared memory object");
+            shm_unlink(p->shm_path);
+            close(p->shm_fd);
+            return;
+        }
+
+        p->buffer = mmap(NULL, p->buffer_size,
+                         PROT_READ | PROT_WRITE, MAP_SHARED, p->shm_fd, 0);
+
+        if (p->buffer == MAP_FAILED) {
+            MP_WARN(vo, "Failed to mmap shared memory object");
+            shm_unlink(p->shm_path);
+            close(p->shm_fd);
+            return;
+        }
+    }
+#endif
+
+    memcpy_pic(p->buffer, p->frame->planes[0], p->width * BYTES_PER_PX,
+               p->height, p->width * BYTES_PER_PX, p->frame->stride[0]);
+
+    if (!p->opts.use_shm)
+        av_base64_encode(p->output, p->output_size, p->buffer, p->buffer_size);
+
+    talloc_free(mpi);
+}
+
+static void flip_page(struct vo *vo)
+{
+    struct priv* p = vo->priv;
+
+    if (p->buffer == NULL)
+        return;
+
+    char *cmd = talloc_asprintf(NULL, TERM_ESC_GOTO_YX, p->top, p->left);
+
+    if (p->opts.use_shm) {
+        cmd = talloc_asprintf_append(cmd, KITTY_ESC_IMG_SHM, p->width, p->height, p->shm_path_b64);
+    } else {
+        if (p->output == NULL) {
+            talloc_free(cmd);
+            return;
+        }
+
+        cmd = talloc_asprintf_append(cmd, KITTY_ESC_IMG, p->width, p->height);
+        for (int offset = 0, noffset;; offset += noffset) {
+            if (offset)
+                cmd = talloc_asprintf_append(cmd, KITTY_ESC_CONTINUE, offset < p->output_size);
+            noffset = MPMIN(4096, p->output_size - offset);
+            cmd = talloc_strndup_append(cmd, p->output + offset, noffset);
+            cmd = talloc_strndup_append(cmd, KITTY_ESC_END, strlen(KITTY_ESC_END));
+
+            if (offset >= p->output_size)
+                break;
+        }
+    }
+
+    write_str(cmd);
+    talloc_free(cmd);
+
+#if HAVE_POSIX
+    if (p->opts.use_shm)
+        close_shm(p);
```

Comment:
is the shm really meant to be closed and reopened every frame?

Reply:
According to the spec, the terminal emulator is supposed to unlink and close it after reading the data.

Reply:
Huh.